Add post about trimming the scheduler. (#1)
CHERIoT-Platform/CHERIoT-Platform.github.io: _posts/2023-11-28-shrinking-the-scheduler.markdown

A futex-like mechanism is now present on all major (non-embedded) operating systems and is even required for C++'s `std::atomic`, which (as of C++20) exposes [`wait`](https://en.cppreference.com/w/cpp/atomic/atomic/wait) and [`notify_one`](https://en.cppreference.com/w/cpp/atomic/atomic/notify_one) / [`notify_all`](https://en.cppreference.com/w/cpp/atomic/atomic/notify_all) methods that expose futex-like behaviour.

A futex is a very general synchronisation primitive.
The wait operation is an atomic compare-and-sleep, which returns immediately if the value in memory is not the expected value.
The notify mechanisms wakes up one or more sleepers.
This works even in a CHERI world with separate permissions.
The wait operation requires read access, the notify operation requires write access, but neither requires both.
After the initial check, the scheduler associates sleepers with an address and so doesn't need to be able to capture the pointer that's passed.

The scheduler has to be in the TCB for availability but ideally it shouldn't be for confidentiality or integrity.
For event groups, this was mostly true.

> A malicious or compromised scheduler could set or clear event notifications causing things to wake spuriously or not, but you could at least mitigate this in by also communicating any key state separately from the event group and using it just to tell you to check something else.

Reviewer: ```suggestion
A malicious or compromised scheduler could set or clear event notifications, causing things to wake spuriously or not, but you could at least mitigate this by also communicating any key state separately from the event group and using it just to tell you to check something else.

```

Message queues complicated this picture.
The scheduler could see all messages in message queues.
Worse, message queues were reachable from the thread structures and so a compromise of the scheduler in one thread could leak or corrupt any data in any in-flight message queues.
This put the scheduler in the TCB for confidentiality and integrity of the contents of message queues.

Futexes can be used to implement a rich set of locks and can also be used to implement message queues and event groups.
This is precisely what [PR139](https://github.com/microsoft/cheriot-rtos/pull/139) is doing.

Message queues are interesting because they have a variety of threat models.
In some cases, they're used for passing messages between threads in the same compartment.
In others, they're expected to be a trusted intermediary between two compartments in different trust domains.

In the first case, both ends trust each other to follow the rules and so we can expose a message-queue library and invoke it without a cross-compartment call.
This makes the overhead of sending and receiving in a message queue lower.

> You need a cross-compartment call only if the queue is transitioning from empty to non-empty, full to non-full, or if there is contention on sending or receiving.

Reviewer: I take it these are the same cases where a futex based lock also requires a compartment call i.e. to sleep / notify? Would be good to be explicit about that.

All other cases are a direct call to a shared library.

> In the second case, we need another compartment to ensure that the message-queue rules are followed.

Reviewer: What rules exactly? Are we talking correctly updating state like head / tail indexes?

The scheduler was doing this, but in the new version there is an option compartment that can fill this rôle.
This compartment is completely unprivileged and compromising it will not impact the scheduler.
It is also almost entirely stateless.
Two different message queues are never visible to it at the same time, all communication between threads happens via message queue objects that are passed in (as sealed capabilities) from callers.
The only state that it owns is the cached thread ID pointer returned from `thread_id_get_pointer()` by the scheduler.

> This means that, if you find a bug in the message queue code (entirely possible!) and force it to crash, it will not affect any other message queues, only yours.

Reviewer: I guess it also means it won't leak data about one queue to another queue.

The wait operation requires read access, the notify operation requires write access, but neither requires both.
After the initial check, the scheduler associates sleepers with an address and so doesn't need to be able to capture the pointer that's passed.

The scheduler has to be in the TCB for availability but ideally it shouldn't be for confidentiality or integrity.
For event groups, this was mostly true.
A malicious or compromised scheduler could set or clear event notifications, causing things to wake spuriously or not, but you could at least mitigate this by also communicating any key state separately from the event group and using it just to tell you to check something else.

Message queues complicated this picture.
The scheduler could see all messages in message queues.
Worse, message queues were reachable from the thread structures and so a compromise of the scheduler in one thread could leak or corrupt any data in any in-flight message queues.
This put the scheduler in the TCB for confidentiality and integrity of the contents of message queues.

Futexes can be used to implement a rich set of locks and can also be used to implement message queues and event groups.

> This is precisely what [PR139](https://github.com/microsoft/cheriot-rtos/pull/139) is doing.

Reviewer 2: Perhaps switch the order of this and the next paragraph?

Message queues are interesting because they have a variety of threat models.
In some cases, they're used for passing messages between threads in the same compartment.
In others, they're expected to be a trusted intermediary between two compartments in different trust domains.

In the first case, both ends trust each other to follow the rules and so we can expose a message-queue library and invoke it without a cross-compartment call.
This makes the overhead of sending and receiving in a message queue lower.

> You need a cross-compartment call only if the queue is transitioning from empty to non-empty, full to non-full, or if there is contention on sending or receiving, the cases where a thread in one endpoint may be sleeping.

Reviewer 2: ```suggestion
This usage needs a cross-compartment call -- to the scheduler -- only if the queue is transitioning from empty to non-empty, full to non-full, or if there is contention on sending or receiving.
That is, the scheduler is only involved in the cases where a thread in one endpoint may be sleeping.
```

All other cases are a direct call to a shared library.

In the second case, we need another compartment to ensure that the message-queue rules are followed, for example not overwriting arbitrary messages and ensuring that the producer and consumer counters are incremented correctly.

> The scheduler was doing this, but in the new version there is an option compartment that can fill this rôle.

Reviewer 2: ```suggestion
The scheduler was doing this, but in the new version there is an optional compartment that can fill this rôle.
```

(And thank you for the diacritic.)

This compartment is completely unprivileged and compromising it will not impact the scheduler.
It is also almost entirely stateless.
Two different message queues are never visible to it at the same time, all communication between threads happens via message queue objects that are passed in (as sealed capabilities) from callers.
The only state that it owns is the cached thread ID pointer returned from `thread_id_get_pointer()` by the scheduler.
This means that, if you find a bug in the message queue code (entirely possible!) and force it to crash, it will not affect any other message queues, only yours.

> In particular, it won't leak data from message queues or corrupt their state.

Reviewer 2: This paragraph seems to skip over what the compartment actually does.  Perhaps just add something simple like "This new compartment encapsulates message queue operations, with message queues exposed via sealed capabilities." after "rôle" (and removing "all communication ... callers", perhaps).

The new code also uses per-queue and per-event-group synchronisation.
The original version in the scheduler ran with interrupts disabled.

> This simplified the code but meant that some operations could delay interrupts for a along time.

Reviewer 2: ```suggestion
This simplified the code but meant that some operations could delay interrupts for a long time.
```
